src/util/lp/vars_equivalence.h

```diff
@@ -0,0 +1,296 @@
+
+  <name>
+
+  Abstract:
+
+  <abstract>
+
+  Author:
+  Nikolaj Bjorner (nbjorner)
+  Lev Nachmanson (levnach)
+
+  Revision History:
+
+
+  --*/
+
+namespace nla {
+
+typedef lp::constraint_index     lpci;
+typedef lp::explanation          expl_set;
+typedef lp::var_index            lpvar;
+struct hash_svector {
+    size_t operator()(const unsigned_vector & v) const {
+        return svector_hash<unsigned_hash>()(v);
+    }
+};
+
+
+
+struct vars_equivalence {
+    
+    struct equiv {
+        lpvar                m_i;
+        lpvar                m_j;
+        rational             m_sign;
+        lpci                 m_c0;
+        lpci                 m_c1;
+
+        equiv(lpvar i, lpvar j, rational const& sign, lpci c0, lpci c1) :
+            m_i(i),
+            m_j(j),
+            m_sign(sign),
+            m_c0(c0),
+            m_c1(c1) {
+            SASSERT(m_i != m_j);
+        }
+    };
+
+    struct node {
+        unsigned            m_parent;   // points to m_equivs
+        svector<unsigned>   m_children;  // point to m_equivs
+        node() : m_parent(-1) {}
+    };
+   
+    //fields
+    
+    // The map from the variables to m_nodes
+    // m_tree is a spanning tree of the graph of equivs represented by m_equivs
+    std::unordered_map<unsigned, node>            m_tree;  
+    // If m_tree[v] == -1 then the variable is a root.
+    // if m_tree[v] is not equal to -1 then m_equivs[m_tree[v]] = (k, v) or (v, k), where k is the parent of v
+    vector<equiv>                                 m_equivs;         // all equivalences extracted from constraints
+    std::unordered_map<rational, unsigned_vector> m_vars_by_abs_values;
+    std::function<rational(lpvar)>                m_vvr;
+    
+
+    // constructor
+    vars_equivalence(std::function<rational(lpvar)> vvr) : m_vvr(vvr) {}
+    
+    void clear() {
+        m_equivs.clear();
+        m_tree.clear();
+        m_vars_by_abs_values.clear(); 
+    }
+
+    const svector<lpvar>& get_vars_with_the_same_abs_val(const rational& v) const {
+        auto it = m_vars_by_abs_values.find(abs(v));
+        SASSERT (it != m_vars_by_abs_values.end());
+        TRACE("nla_solver", tout << "size of same_abs_vals = " << it->second.size(); );
+        return it->second; 
+    } 
+
+    // j itself is also included
+    svector<lpvar> eq_vars(lpvar j) const {
+        svector<lpvar> r = path_to_root(j);
+        svector<lpvar> ch = children(j);
+        for (lpvar k : ch) {
+            r.push_back(k);
+        }
+        r.push_back(j);
+        return r;
+    } 
+
+    svector<lpvar> children(lpvar j) const {
+        svector<lpvar> r;
+        auto it = m_tree.find(j);
+        if (it == m_tree.end())
+            return r;
+
+        const node& n = it->second;
+        for (unsigned e_k: n.m_children) {
+            const equiv & e = m_equivs[e_k];
+            lpvar oj = e.m_i == j? e.m_j : e.m_i;
+            r.push_back(oj);
+            for (lpvar k :  children(oj)) {
```

Comment:
is this shallow/non-shared?

Reply:
Sorry, I do not understand the question.